### PR TITLE
add thiserror to sozu-command-lib

### PR DESCRIPTION
**Pull request by Keksoj**
Using [anyhow](https://docs.rs/anyhow/latest/anyhow/index.html) in Sōzu has proven valuable to concatenate context accross function, and display the whole context all in one line for optimal readability.

This readability is only an advantage to humans, though. 

We aim to make errors more usable for machines, with proper error enums to be pattern-matched and so on.

### CODE REVIEW

**Comment by Keksoj** [comment]
This is to solve this issue:

- #910

File: command/src/channel.rs [hunk 1]
```diff
@@ -12,13 +12,42 @@ use std::{
...
+    #[error("invalid utf-8 encoding in command message, ignoring: {0}")]
+    Utf8Error(String),
+    #[error("Error deserializing message")]
+    SerdeError(serde_json::error::Error),
+    #[error("Could not change the blocking status ef the unix stream with file descriptor {fd}: {error}")]
+    BlockingStatusError { fd: i32, error: String },
+    #[error("Connection error, continuing")]
+    ConnectionError,
+    #[error("Could not connect to socket with path {path}: {error}")]
+    FromPathError { path: String, error: String },
```
Comment (FlorentinDUBOIS) [inline command/src/channel.rs:48]: Should be a `Connection` error

File: command/src/channel.rs [hunk 2]
```diff
@@ -12,13 +12,42 @@ use std::{
...
 use mio::{event::Source, net::UnixStream as MioUnixStream};
 use serde::{de::DeserializeOwned, ser::Serialize};
 use serde_json;
 
 use crate::{buffer::growable::Buffer, ready::Ready};
 
+#[derive(thiserror::Error, Debug)]
+pub enum ChannelError {
+    #[error("io read error")]
+    IoError(std::io::Error),
```
Comment (FlorentinDUBOIS) [inline command/src/channel.rs:24]: Too generic

File: command/src/channel.rs [hunk 3]
```diff
@@ -12,13 +12,42 @@ use std::{
...
+    #[error("no byte written on the channel")]
+    NoByteWritten,
+    #[error("no byte left to read on the channel")]
+    NoByteToRead,
+    #[error("message too large for the capacity of the back fuffer ({0}. Consider increasing the back buffer size")]
+    MessageTooLarge(usize),
+    #[error("channel could not write on the back buffer")]
+    WriteError(std::io::Error),
+    #[error("channel buffer is full, cannot grow more, ignoring")]
```
Comment (FlorentinDUBOIS) [inline command/src/channel.rs:33]:
```suggestion
    #[error("channel buffer is full, cannot grow more")]
```

File: command/src/channel.rs [hunk 4]
```diff
@@ -12,13 +12,42 @@ use std::{
...
+    BufferFull,
+    #[error("Timeout is reached: {0:?}")]
+    TimeoutReached(Duration),
+    #[error("Could not read anything on the channel")]
+    NothingRead,
+    #[error("invalid utf-8 encoding in command message, ignoring: {0}")]
+    Utf8Error(String),
```
Comment (FlorentinDUBOIS) [inline command/src/channel.rs:40]:
```suggestion
    InvalidCharset(String),
```

File: command/src/channel.rs [hunk 5]
```diff
@@ -12,13 +12,42 @@ use std::{
...
+    #[error("Error deserializing message")]
+    SerdeError(serde_json::error::Error),
+    #[error("Could not change the blocking status ef the unix stream with file descriptor {fd}: {error}")]
+    BlockingStatusError { fd: i32, error: String },
+    #[error("Connection error, continuing")]
```
Comment (FlorentinDUBOIS) [inline command/src/channel.rs:45]:
```suggestion
    #[error("Connection error")]
```

File: command/src/config.rs [hunk 6]
```diff
@@ -155,6 +154,55 @@ pub const DEFAULT_COMMAND_BUFFER_SIZE: usize = 1_000_000;
 /// maximum size of the buffer for the channels, in bytes. (2 MB)
 pub const DEFAULT_MAX_COMMAND_BUFFER_SIZE: usize = 2_000_000;
 
+#[derive(thiserror::Error, Debug)]
+pub enum ConfigError {
+    #[error("Could not parse socket address {address}: {error}")]
+    AddressParsingError { address: String, error: String },
```
Comment (FlorentinDUBOIS) [inline command/src/config.rs:160]:
```suggestion
    ParseSocketAddr { address: String, error: String },
```

File: command/src/scm_socket.rs [hunk 7]
```diff
@@ -8,14 +8,27 @@ use std::{
...
+#[derive(thiserror::Error, Debug)]
+pub enum ScmSocketError {
+    #[error("could not set the blocking status of the unix stream to {blocking}: {error}")]
+    SetBlocking { blocking: bool, error: String },
+    #[error("could not send message per SCM socket: {0}")]
+    SendError(String),
+    #[error("could not send message per SCM socket: {0}")]
+    ReceiveError(String),
+    #[error("could not parse utf8 from buffer: {0}")]
+    Utf8Error(String),
```
Comment (FlorentinDUBOIS) [inline command/src/scm_socket.rs:27]:
```suggestion
    InvalidCharset(String),
```

File: command/src/state.rs [hunk 8]
```diff
@@ -29,6 +27,48 @@ use crate::{
...
+    #[error("Request came in empty")]
+    EmptyRequest,
+    #[error("dispatching this request did not bring any change to the state")]
+    NoChange,
+    #[error("State can not handle this request")]
+    UndispatchableRequest,
+    #[error("Did not find cluster {0}")]
+    ClusterNotFound(String),
+    #[error("Listener exists already on address {0}")]
+    ListenerExistsAlready(String),
```
Comment (FlorentinDUBOIS) [inline command/src/state.rs:41]:
```suggestion
    Exists(String),
```

File: command/src/state.rs [hunk 9]
```diff
@@ -29,6 +27,48 @@ use crate::{
...
+    #[error("Wrong request: {0}")]
+    WrongRequest(String),
+    #[error("No listener to remove at address {0}")]
+    NoListenerToRemove(String),
```
Comment (FlorentinDUBOIS) [inline command/src/state.rs:45]:
```suggestion
    NotFound(String),
```

File: command/src/state.rs [hunk 10]
```diff
@@ -29,6 +27,48 @@ use crate::{
...
+#[derive(thiserror::Error, Debug)]
+pub enum StateError {
+    #[error("Request came in empty")]
+    EmptyRequest,
+    #[error("dispatching this request did not bring any change to the state")]
+    NoChange,
+    #[error("State can not handle this request")]
+    UndispatchableRequest,
+    #[error("Did not find cluster {0}")]
+    ClusterNotFound(String),
```
Comment (FlorentinDUBOIS) [inline command/src/state.rs:39]:
```suggestion
    NotFound(String),
```

**Review by FlorentinDUBOIS — CHANGES_REQUESTED** [review]
Hey :wave: , I started the review, but I do not want to create a lot of comments. 
Could we avoid to
- Create items on error that ends with `Error`, it is redundant (aka `ChannerError::FooError`)
- Have generic message error like `IoError(std::io::Error)` and with a message like `read error`
- Create redendundant entries like (ClusterNotFound, ListenerNotFound and so on..), but instead have an error like `NotFound { kind: AnEnum, id: String }`

**Comment by Keksoj** [comment]
thanks @FlorentinDUBOIS , I removed all `...Error` endings in variants, it's more legible :+1: and I merged redundant error types in `NotFound` or `NoMetrics` variants.

File: command/src/config.rs [hunk 11]
```diff
@@ -155,6 +154,55 @@ pub const DEFAULT_COMMAND_BUFFER_SIZE: usize = 1_000_000;
...
+    FileRead {
+        path_to_read: String,
+        io_error: String,
+    },
+    #[error("the listener on {0} has incompatible options: it cannot use the expect proxy protocol and have a public_address field at the same time",)]
+    IncompatiblePublicAddress(String),
+    #[error("all the listeners for cluster {0} should have the same expect_proxy option")]
+    InconsistentExpectProxyOption(String),
+    #[error("Invalid '{0}' field for a TCP frontend")]
+    InvalidField(String),
```
Comment (FlorentinDUBOIS) [inline command/src/config.rs:178]: Should be an enum with error `Invalid` and kind `Field` and `Path`

File: command/src/config.rs [hunk 12]
```diff
@@ -155,6 +154,55 @@ pub const DEFAULT_COMMAND_BUFFER_SIZE: usize = 1_000_000;
...
+    #[error("invalid path {0:?}")]
+    InvalidPath(PathBuf),
+    #[error("listening address {0} is already used in the configuration")]
+    ListenerAddressAlreadyInUse(String),
+    #[error("Invalid '{0}' field to create a frontend")]
+    MissingField(String),
```
Comment (FlorentinDUBOIS) [inline command/src/config.rs:184]: Should be an enum with error `NotFound` with kind `Field(string)`, `Protocol` and `SavedState`

File: command/src/state.rs [hunk 13]
```diff
@@ -29,6 +27,47 @@ use crate::{
...
+#[derive(thiserror::Error, Debug)]
+pub enum StateError {
+    #[error("Request came in empty")]
+    EmptyRequest,
+    #[error("dispatching this request did not bring any change to the state")]
+    NoChange,
+    #[error("State can not handle this request")]
+    UndispatchableRequest,
+    #[error("Did not find object with address or id '{0}'")]
+    NotFound(String),
```
Comment (FlorentinDUBOIS) [inline command/src/state.rs:50]:
```suggestion
    NotFound(ObjectKind, string),
```
Am I missing a thing ?

File: command/src/state.rs [hunk 14]
```diff
@@ -29,6 +27,47 @@ use crate::{
...
+    #[error("{kind:?} {id} already exists")]
+    Exists { kind: ObjectKind, id: String },
+    #[error("Wrong request: {0}")]
+    WrongRequest(String),
+    #[error("No listener to remove at address {0}")]
+    NoListenerToRemove(String),
```
Comment (FlorentinDUBOIS) [inline command/src/state.rs:56]: redundant with `NoChange`

File: lib/src/lib.rs [hunk 15]
```diff
@@ -629,6 +639,36 @@ pub enum BackendConnectAction {
     Replace,
 }
 
+#[derive(thiserror::Error, Debug)]
+pub enum BackendConnectionError {
+    #[error("Found no TCP cluster")]
+    FoundNoTcpCluster,
```
Comment (FlorentinDUBOIS) [inline lib/src/lib.rs:645]:
```suggestion
    NotFound(ObjectKind),
```

File: lib/src/lib.rs [hunk 16]
```diff
@@ -639,6 +679,76 @@ pub enum AcceptError {
...
 
+/// returned by the HTTP, HTTPS and TCP listeners
+#[derive(thiserror::Error, Debug)]
+pub enum ListenerError {
+    #[error("failed to acquire the lock, {0}")]
+    Lock(String),
+    #[error("failed to handle certificate request, got a resolver error, {0}")]
+    Resolver(GenericCertificateResolverError),
+    #[error("failed to parse pem, {0}")]
+    PemParse(String),
```
Comment (FlorentinDUBOIS) [inline lib/src/lib.rs:690]:
```suggestion
    InvalidPem(String),
```

File: lib/src/lib.rs [hunk 17]
```diff
@@ -639,6 +679,76 @@ pub enum AcceptError {
...
+    #[error("failed to build rustls context, {0}")]
+    BuildRustls(String),
+    #[error("Wrong socket address")]
+    SocketParse { address: String, error: String },
```
Comment (FlorentinDUBOIS) [inline lib/src/lib.rs:694]:
```suggestion
    InvalidSocketAddress { address: String, error: String },
```

File: lib/src/metrics/mod.rs [hunk 18]
```diff
@@ -23,6 +22,26 @@ thread_local! {
   pub static METRICS: RefCell<Aggregator> = RefCell::new(Aggregator::new(String::from("sozu")));
 }
 
+#[derive(thiserror::Error, Debug)]
+pub enum MetricError {
+    #[error("Could not parse udp address {address}: {error}")]
+    WrongUdpAddress { address: String, error: String },
```
Comment (FlorentinDUBOIS) [inline lib/src/metrics/mod.rs:28]:
```suggestion
    InvalidUdpAddress { address: String, error: String },
```

File: lib/src/router/mod.rs [hunk 19]
```diff
@@ -14,6 +13,18 @@ use sozu_command::{
 
 use self::pattern_trie::TrieNode;
 
+#[derive(thiserror::Error, Debug)]
+pub enum RouterError {
+    #[error("Could not parse rule from frontend path {0:?}")]
+    PathRuleParse(String),
```
Comment (FlorentinDUBOIS) [inline lib/src/router/mod.rs:19]:
```suggestion
    InvalidPathRule(String),
```

File: lib/src/router/mod.rs [hunk 20]
```diff
@@ -14,6 +13,18 @@ use sozu_command::{
...
+    #[error("parsing hostname {hostname} failed")]
+    DomainParse { hostname: String },
```
Comment (FlorentinDUBOIS) [inline lib/src/router/mod.rs:21]:
```suggestion
    InvalidDomain { hostname: String },
```

File: lib/src/socket.rs [hunk 21]
```diff
@@ -3,11 +3,28 @@ use std::{
...
+    #[error("could not set socket to nonblocking: {0}")]
+    SetNonBlocking(String),
+    #[error("could not set reuse address: {0}")]
+    SetReuseAddress(String),
+    #[error("could not set reuse address: {0}")]
+    SetReusePort(String),
+    #[error("Could not create socket: {0}")]
+    SocketCreationError(String),
+    #[error("Wrong socket address '{address}': {error}")]
+    WrongAddress { address: String, error: String },
```
Comment (FlorentinDUBOIS) [inline lib/src/socket.rs:25]:
```suggestion
    InvalidSocketAddress { address: String, error: String },
```

File: lib/src/tls.rs [hunk 22]
```diff
@@ -153,15 +153,15 @@ impl Clone for ParsedCertificateAndKey {
 #[derive(thiserror::Error, Clone, Debug)]
 pub enum GenericCertificateResolverError {
     #[error("failed to parse certificate common name, {0}")]
-    CommonNameParseError(String),
+    CommonNameParse(String),
```
Comment (FlorentinDUBOIS) [inline lib/src/tls.rs:156]: Could you rename to 
```suggestion
    InvalidCommonName(String),
```

File: lib/src/tls.rs [hunk 23]
```diff
@@ -153,15 +153,15 @@ impl Clone for ParsedCertificateAndKey {
...
     #[error("failed to parse pem certificate, {0}")]
-    PemParseError(String),
+    PemParse(String),
```
Comment (FlorentinDUBOIS) [inline lib/src/tls.rs:158]:
```suggestion
    InvalidPem(String),
```

File: lib/src/tls.rs [hunk 24]
```diff
@@ -153,15 +153,15 @@ impl Clone for ParsedCertificateAndKey {
...
     #[error("failed to parse der certificate, {0}")]
-    DerParseError(String),
+    DerParse(String),
```
Comment (FlorentinDUBOIS) [inline lib/src/tls.rs:160]:
```suggestion
    InvalidDer(String),
```

**Review by FlorentinDUBOIS — CHANGES_REQUESTED** [review]
Nice works ! There are some stuffs to do, but we are quite close to merge it.

File: command/src/channel.rs [hunk 25]
```diff
@@ -39,9 +66,9 @@ impl<Tx: Debug + Serialize, Rx: Debug + DeserializeOwned> Channel<Tx, Rx> {
         path: &str,
         buffer_size: usize,
         max_buffer_size: usize,
-    ) -> anyhow::Result<Channel<Tx, Rx>> {
+    ) -> Result<Channel<Tx, Rx>, ChannelError> {
         let unix_stream = MioUnixStream::connect(path)
-            .with_context(|| format!("Could not connect to socket with path {path}"))?;
+            .map_err(|io_error| ChannelError::Connection(Some(io_error.to_string())))?;
```
[the remark this answers is not in the document]
Comment (Keksoj) [reply command/src/channel.rs:71]: Well, I find writing the error type in the clojure makes it very legible.

Or do you mean removing the `.to_string()` and implementing `From<IoError>` ?

File: command/src/channel.rs [hunk 26]
```diff
@@ -12,13 +12,40 @@ use std::{
...
+    #[error("Could not read anything on the channel")]
+    NothingRead,
+    #[error("invalid char set in command message, ignoring: {0}")]
+    InvalidCharSet(String),
+    #[error("Error deserializing message")]
+    Serde(serde_json::error::Error),
+    #[error("Could not change the blocking status ef the unix stream with file descriptor {fd}: {error}")]
+    BlockingStatus { fd: i32, error: String },
+    #[error("Connection error: {0:?}")]
+    Connection(Option<String>),
```
[the remark this answers is not in the document]
Comment (Keksoj) [reply command/src/channel.rs:46]: Sure !

File: command/src/channel.rs [hunk 27]
```diff
@@ -153,17 +179,17 @@ impl<Tx: Debug + Serialize, Rx: Debug + DeserializeOwned> Channel<Tx, Rx> {
...
-                    bail!("Error with the socket, reading on it return 0");
+                    return Err(ChannelError::NoByteToRead);
                 }
                 Err(read_error) => match read_error.kind() {
                     ErrorKind::WouldBlock => {
                         self.readiness.remove(Ready::READABLE);
                         break;
                     }
-                    other_error => {
+                    _other_error => {
```
[the remark this answers is not in the document]
Comment (Keksoj) [reply command/src/channel.rs:189]: very well if you insist

File: command/src/state.rs (hunk 13 above)
Comment (Keksoj) [reply command/src/state.rs:50]: I hope not, I'll take that into account

**Comment by Keksoj** [comment]
this is ready for review BTW

**Review by FlorentinDUBOIS — APPROVED** [review]

**Pull request merged**